firestore/api/SolutionCounter/Program.cs

```diff
@@ -0,0 +1,209 @@
+    /// Defines the <see cref="DistributedCounter" />
+    /// </summary>
+    public class DistributedCounter
+    { 
+        public static string Usage = @"Usage:
+C:\> dotnet run command YOUR_PROJECT_ID
+
+Where command is one of
+    run-distributed-counter
+";
+        /// <summary>
+        /// Gets or sets the FireStoreDb
+        /// </summary>
+        public static FirestoreDb FirestoreDb { get; set; }
+
+        /// <summary>
+        /// Gets or sets the ShardsCounter
+        /// </summary>
+        public static Counter ShardsCounter { get; set; }
+
+        /// <summary>
+        /// The InitFireStoreDb
+        /// </summary>
+        public static void InitFirestoreDb(string project)
+        {
+            var projectId = project;
+            if (string.IsNullOrEmpty(projectId))
+                throw new ArgumentException("ArgumentException: failed read Firestore Project ID");
+
+            FirestoreDb = FirestoreDb.Create(projectId: projectId);
+            Console.WriteLine($"Created Cloud Firestore client with project ID: {project}");
+        }
+
+        /// <summary>
+        /// The InitShardsCounter
+        /// </summary>
+        /// <param name="numShards">The numShards<see cref="int"/></param>
+        public static void InitShardsCounter(int numShards)
+        {
+            // Initialize whole global count of shards
+            // in ShardsCounter property 
+            ShardsCounter = new Counter
+            {
+                NumShards = numShards
+            };
+        }
+
+        // [START fs_create_counter]
+
+        /// <summary>
+        /// The InitCounterAsync creates a given number of shards as
+        /// sub collection of specified document.
+        /// </summary>
+        /// <param name="docRef">The docRef<see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="Task"/></returns>
+        public static async Task InitCounterAsync(DocumentReference docRef)
+        {
+            CollectionReference colRef = docRef.Collection("shards");
+
+            // Initialize each shard with count=0
+            for (var i = 0; i < ShardsCounter.NumShards; i++)
+            {
+                await colRef.Document(i.ToString()).SetAsync(new Shard { Count = 0 });
+            }
+        }
+
+        // [END fs_create_counter]
+
+        // [START fs_increment_counter]
+
+        /// <summary>
+        /// The IncrementCounterAsync increments a randomly picked shard.
+        /// </summary>
+        /// <param name="docRef">The docRef<see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="Task"/></returns>
+        public static async Task IncrementCounterAsync(DocumentReference docRef)
+        {
+            var rand = new Random();
+            var docId = rand.Next(0, ShardsCounter.NumShards);
+            var shardRef = docRef.Collection("shards").Document(docId.ToString());
+            await shardRef.UpdateAsync("Count", FieldValue.Increment(1));
```

Comment:
I'm not that familiar with the Firestore API, but how does UpdateAsync manages concurrent modifications? If it doesn't fail on concurrent modifcations, then in my opinion this whole example is not a good example.